### PR TITLE
Bug fix corporation expressions

### PR DESCRIPTION
**Pull request by ltoni90**
Fixed bug that caused an error when trying to set a dynamic expression to sell materials or products (e.g. PROD-500). [See Issue](https://github.com/danielyxie/bitburner/issues/2719)

**Changes**
- Added property "maxsll" to Material.ts and Product.ts
- The result of the calculation of maxSell will now be written to the property "maxsll" of the corresponding material or product
- Temporary quantity evaluation does not use "1" anymore. Instead it uses the effective production rate respectively the maximum amount of the product or material that can be sold.

**How it was tested**
Entered expressions where the evaluation leads to negative result and error message was displayed as expected.

### CODE REVIEW

File: src/Corporation/Product.ts [hunk 1]
```diff
@@ -96,6 +96,8 @@ export class Product {
   marketTa2 = false;
   marketTa2Price: IMap<number> = createCityMap<number>(0);
 
+  // Determines the maximum amount of this product that can be sold in one market cycle
+  maxsll = 0
```
Comment (ChrissiQ) [inline src/Corporation/Product.ts:100]:
```suggestion
  maxsll = 0;
```

**Comment by ltoni90** [comment]
Is there any reason why this PR won't get approved? Can I do something to speed this up?

**Comment by phyzical** [comment]
sadly you will just have to be patient :)

**Comment by Penumbra69** [comment]
Sadly this is a game breaker.

**Comment by phyzical** [comment]
@Penumbra69 passive aggressive responses like this wont help anything..

**Review by phyzical — COMMENTED** [review]
LGTM, once theres, a test script and some sort of gif showing how it was tested i can approve.

(need to consider things like too small numbers, too big, -ves ect.

**Comment by Penumbra69** [comment]
> @Penumbra69 passive aggressive responses like this wont help anything..

Neither is being told to "sadly" wait for a game breaking bug to be fixed when a PR is offered to fix it without any advice on timing or priority.

If your words, which I just echoed back, were seen as passive aggressive - what would that mean about yours?
Aggressively passive?! :D 

Anyhow - thanks for looking at the PR, I hope this bug fix makes it through the gauntlet soon so we can all re-enjoy this part of the game!

Thanks,
P

**Comment by phyzical** [comment]
yes... bark and you will get barked back at

**Comment by ltoni90** [comment]
Hey @phyzical ,

At the time where I opened this pull request, my corporation wasn't as advanced as of now and while creating the test GIFs I stumbled upon another problem regarding big floating numbers.

This are my current stats:
grafik

Which results in maxsll of 1969593915067574800. At this point expressions like MAX-(MAX+0.0001) or even MAX-(MAX+10) will eval to 0 and thus validating as true instead of false. I know it's a silly expression and also a highly advanced corporation I'm talking about here, but nonetheless the validation logic wouldn't work in this case. If you say that this is a no-go, I'll need more time to implement a way that satisfies such edgy cases. What do you think?

**Comment by phyzical** [comment]
@ltoni90 ah damn this is why i like the idea for forcing a gif :) you end up wanting to test more than you thought and find those edge cases that just cant bee seen otherwise.

Given the time it will take to get into prod/dev anyway i think it would be reasonable to have another think about how to solve it. (simpler the better :) ) i feel like if there is a way to achieve this without storing at against the mat itself also would be better, just feel like it may be exploitable 😆 

there no rush :)

if i was to guess its toString messing with you, maybe theres some helper function in the codebase exactly for this  "convert number to string safely"

@pigalot do you have any suggestions (incase this just isnt the right way to solve this)

**Comment by pigalot** [comment]
I would not bother trying to solve for very stupid edge cases like "MAX-(MAX+0.0001) or even MAX-(MAX+10)".
If they evaluate to 0 then that's fine it will just set it to 0, if the user wants too shoot them self in the foot then let them as long as it prevents negative numbers it doesn't matter.

**Comment by ltoni90** [comment]
@pigalot @phyzical 
sounds reasonable for me to ignore some edgy cases. The only way I see right now to handle big numbers would be to use shifted BigInts. But as we need to evaluate mathematical expressions it would be necessary to compile the expression in some sort of math expression tree as point and dash operations have to be calculated in a special way (like [here](https://stackoverflow.com/a/66939244)) when working with shifted BigInts. And that's way to exaggerated for this use case....

@phyzical 
regarding your concerns about the possibilty of exploitation. The additional property on Material.ts and Product.ts is only used for the validation of a sell expression and won't be accessed afterwards (it only gets updated in each market cycle). And before selling the actual material or product the expression is getting evaluated again anyway. If the result should be negative than it will be set to 0 (see [here](https://github.com/danielyxie/bitburner/blob/ee887de151f00c508e4f37ba94885ceeaffa36fd/src/Corporation/Industry.ts#L870) or [here](https://github.com/danielyxie/bitburner/blob/ee887de151f00c508e4f37ba94885ceeaffa36fd/src/Corporation/Industry.ts#L1228)). Because of that I don't really believe that the property can be exploited in any way. I've also looked into some alternatives like calculating maxsll in Actions.ts, but that's not possible as some variables used to calculate maxsll are only available during [processMaterials()](https://github.com/danielyxie/bitburner/blob/ee887de151f00c508e4f37ba94885ceeaffa36fd/src/Corporation/Industry.ts#L523) or [processProdcuts()](https://github.com/danielyxie/bitburner/blob/ee887de151f00c508e4f37ba94885ceeaffa36fd/src/Corporation/Industry.ts#L993). So I think that the simplest solution to this problem is to create the maxsll property on Material or Product instance.

If that's okay for you I will create the test GIFs and post them here.

**Comment by phyzical** [comment]
👍

**Comment by ltoni90** [comment]
@phyzical here you go:

Test 1
Division: Real Estate
Type: Material - Real Estate
Expression: PROD-100
Expected Result: OK
Result: OK
Test_1


Test 2
Division: Real Estate
Type: Material - Real Estate
Expression: PROD-60000
Expected Result: FAIL
Result: FAIL
Test_2

Test 3
Division: Real Estate
Type: Product
Expression: PROD-100
Expected Result: OK
Result: OK
Test_3


Test 4
Division: Real Estate
Type: Product
Expression: PROD-31000
Expected Result: FAIL
Result: FAIL
Test_4

**Comment by phyzical** [comment]
looking good GJ thanks for the gifs, i wonder if we should improve the error message too. also how about api support?
